src/backend/distributed/commands/local_multi_copy.c

```diff
@@ -0,0 +1,238 @@
+ */
+static bool
+ShouldSendCopyNow(StringInfo buffer)
+{
+	return buffer->len > LOCAL_COPY_SWITCH_OVER_THRESHOLD;
+}
+
+
+/*
+ * DoLocalCopy finds the shard table from the distributed relation id, and copies the given
+ * buffer into the shard.
+ */
+static void
+DoLocalCopy(StringInfo buffer, Oid relationId, int64 shardId, CopyStmt *copyStatement)
+{
+	localCopyBuffer = buffer;
+
+	Oid shardOid = GetShardOid(relationId, shardId);
+	Relation shard = heap_open(shardOid, RowExclusiveLock);
+	Relation copiedShard = CreateCopiedShard(copyStatement->relation, shard);
+	ParseState *pState = make_parsestate(NULL);
+
+	/* p_rtable of pState is set so that we can check constraints. */
+	pState->p_rtable = CreateRangeTable(copiedShard, ACL_INSERT);
+
+	CopyState cstate = BeginCopyFrom(pState, copiedShard, NULL, false,
+									 ReadFromLocalBufferCallback,
+									 copyStatement->attlist, copyStatement->options);
+	CopyFrom(cstate);
+	EndCopyFrom(cstate);
+
+	free_parsestate(pState);
+	FreeStringInfo(buffer);
+	buffer = makeStringInfo();
+	heap_close(shard, NoLock);
+}
+
+
+/*
+ * ShouldAddBinaryHeaders returns true if the given buffer
+ * is empty and the format is binary.
+ */
+static bool
+ShouldAddBinaryHeaders(StringInfo buffer, bool isBinary)
+{
+	if (!isBinary)
+	{
+		return false;
+	}
+	return buffer->len == 0;
+}
+
+
+/*
+ * CreateCopiedShard clones deep copies the necessary fields of the given
+ * relation.
+ */
+Relation
+CreateCopiedShard(RangeVar *distributedRel, Relation shard)
+{
+	TupleDesc tupleDescriptor = RelationGetDescr(shard);
+
+	Relation copiedDistributedRelation = (Relation) palloc(sizeof(RelationData));
+	Form_pg_class copiedDistributedRelationTuple = (Form_pg_class) palloc(
+		CLASS_TUPLE_SIZE);
+
+	*copiedDistributedRelation = *shard;
+	*copiedDistributedRelationTuple = *shard->rd_rel;
+
+	copiedDistributedRelation->rd_rel = copiedDistributedRelationTuple;
+	copiedDistributedRelation->rd_att = CreateTupleDescCopyConstr(tupleDescriptor);
+
+	Oid tableId = RangeVarGetRelid(distributedRel, NoLock, false);
+
+	/*
+	 * BeginCopyFrom opens all partitions of given partitioned table with relation_open
+	 * and it expects its caller to close those relations. We do not have direct access
+	 * to opened relations, thus we are changing relkind of partitioned tables so that
+	 * Postgres will treat those tables as regular relations and will not open its
+	 * partitions.
+	 */
+	if (PartitionedTable(tableId))
+	{
+		copiedDistributedRelationTuple->relkind = RELKIND_RELATION;
+	}
+	return copiedDistributedRelation;
+}
+
+
+/*
+ * ReadFromLocalBufferCallback is the copy callback.
+ * It always tries to copy maxread bytes.
+ */
+static int
+ReadFromLocalBufferCallback(void *outbuf, int minread, int maxread)
+{
+	int bytesread = 0;
+	int avail = localCopyBuffer->len - localCopyBuffer->cursor;
+	int bytesToRead = Min(avail, maxread);
+	if (bytesToRead > 0)
+	{
+		memcpy_s(outbuf, bytesToRead + strlen((char *) outbuf),
```

Comment:
when we write to the output buffer, if it needs to expand it size, it could reach to even more than `bytesToRead + strlen((char *) outbuf`, So we might need to put something safer here.

Reply:
I guess here we should use 2 * LOCAL_COPY_FLUSH_THRESHOLD

Reply:
does this still need to be resolved? the strlen(outbuf) seems a bit weird

Reply:
`bytesToRead + strlen((char *) outbuf` is the maximum size after doing the copy operation the the buffer.

So the reason I am saying this might not be correct is, say, that the `outbuf` size is X and we are writing Y bytes, then the new buffer size could be more than X + Y if the buffer expands its size.

Though looking at the source code for mem_cpy_s it doesnt seem like it will be an issue.

Reply:
but how can strlen(outbuf) be correct if it does not contain a string?

Reply:
I think just using `bytesToRead` would be fine, since we're guaranteed that `maxread` will not exceed the size of `outbuf`.